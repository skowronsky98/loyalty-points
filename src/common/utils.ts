import crypto from 'crypto';

export const setDefault = <T>(
  value: T,
  defaultValue: Exclude<T, undefined>,
): Exclude<T, undefined> =>
  value === undefined ? defaultValue : (value as Exclude<T, undefined>);

export const hashEvent = (input: { eventName: string, sequence: number }): string => {
  const { eventName, sequence } = input;
  return crypto
    .createHash('sha256')
    .update(`${eventName}-${sequence}`)
    .digest('hex');
};
