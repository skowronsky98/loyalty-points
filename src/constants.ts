import { UniqueConstraintViolationException } from "@mikro-orm/postgresql";

export enum EventName {
  CUSTOMER_CREATED = "CustomerCreated",
  CUSTOMER_DELETED = "CustomerDeleted",
  ORDER_CANCELED = "OrderCanceled",
  ORDER_PLACED = "OrderPlaced",
  ORDER_RETURNED = "OrderReturned",
}

export enum EntityName {
  CUSTOMER = "Customer",
  ORDER = "Order",
}

export const catchWebhookEventsDbErrors = (err: unknown) => {
  if (
    err instanceof UniqueConstraintViolationException &&
    err.message.includes('events_history_event_hash_unique')
  ) {
    return;
  }
  throw err;
};
