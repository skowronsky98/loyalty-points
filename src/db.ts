import { EntityManager, MikroORM, Options } from "@mikro-orm/postgresql";
import mikroOrmConfig from "./mikro-orm.config.js";
import { LoyaltyPointsEntity } from "./entitites/loyalty-points.entity.js";
import { LoyaltyPointsRepository } from "./repos/loyalty-points.repo.js";
import { LoyaltyPointsDeptRepository } from "./repos/loyalty-points-dept.repo.js";
import { LoyaltyPointsDeptEntity } from "./entitites/loyalty-points-dept.entity.js";

export interface Services {
  orm: MikroORM;
  em: EntityManager;
  loyaltyPointsRepo: LoyaltyPointsRepository;
  loyaltyPointsDeptRepo: LoyaltyPointsDeptRepository;
}

let cache: Services;

export async function initORM(options?: Options): Promise<Services> {
  if (cache) {
    return cache;
  }

  const orm = await MikroORM.init({
    ...mikroOrmConfig,
    ...options,
  });

  return cache = {
    orm,
    em: orm.em,
    loyaltyPointsRepo: orm.em.getRepository(LoyaltyPointsEntity),
    loyaltyPointsDeptRepo: orm.em.getRepository(LoyaltyPointsDeptEntity),
  };
}
