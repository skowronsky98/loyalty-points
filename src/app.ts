import Fastify from 'fastify'
import { ENV } from './env/env.js'
import { initORM } from './db.js'
import { RequestContext } from '@mikro-orm/postgresql'
import { registerWebhookEvents } from './endpoints/webhook-events.endpoint.js'
import { registerPointsRoutes } from './endpoints/get-loyalty-points.endpoint.js'
import { registerConsumePoints } from './endpoints/consume-loyalty-points.endpoint.js'
import {
  jsonSchemaTransform,
  serializerCompiler,
  validatorCompiler,
  isResponseSerializationError,
  hasZodFastifySchemaValidationErrors,
  ZodTypeProvider,
} from 'fastify-type-provider-zod';
import fastifySwagger from '@fastify/swagger'
import fastifySwaggerUI from '@fastify/swagger-ui';

export const createApp = async () => {

  const db = await initORM();

  const app = Fastify().withTypeProvider<ZodTypeProvider>();

  app.setValidatorCompiler(validatorCompiler);
  app.setSerializerCompiler(serializerCompiler);

  app.register(fastifySwagger, {
    openapi: {
      info: {
        title: 'Loyalty Points Service',
        description: 'Sample backend service',
        version: '1.0.0',
      },
      servers: [],
    },
    transform: jsonSchemaTransform,
  });

  app.register(fastifySwaggerUI, {
    routePrefix: '/documentation',
  });


  app.addHook('onRequest', (request, reply, done) => {
    RequestContext.create(db.em, done);
  });

  app.setErrorHandler((error, request, reply) => {

    if (hasZodFastifySchemaValidationErrors(error)) {
      return reply.status(400).send({ error: 'Bad Request', statusCode: 400 });
    }

    if (isResponseSerializationError(error)) {
      return reply.code(500).send({
        error: 'Internal Server Error',
        statusCode: 500,
      })
    }

    app.log.error(error);

    return reply.code(500).send({
      error: 'Internal Server Error',
      statusCode: 500,
    })
  });

  app.addHook('onClose', async () => {
    await db.orm.close();
  });

  app.register(registerPointsRoutes);
  app.register(registerConsumePoints);
  app.register(registerWebhookEvents);

  await app.ready();
  await app.listen({ port: ENV.APP_PORT });

  console.log(`Documentation running at http://localhost:3000/documentation`);

  return app;
}

try {
  await createApp();
} catch (error) {
  console.error(error);
  process.exit(1);
}
