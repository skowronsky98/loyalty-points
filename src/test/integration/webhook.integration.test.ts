import { FastifyInstance } from 'fastify'
import { afterAll, beforeAll, beforeEach, describe, expect, it } from 'vitest'
import supertest from 'supertest'
import { initORM, Services } from '../../db.js'
import { createApp } from '../../app.js'
import { clearDatabase } from '../../test-utils/test.utils.js'
import { EntityManager } from '@mikro-orm/postgresql'
import { LoyaltyPointsEntity } from '../../entitites/loyalty-points.entity.js'
import { randomUUID } from 'node:crypto'
import { addDays } from 'date-fns'
import { hashEvent } from '../../common/utils.js'
import { EventsHistoryEntity } from '../../entitites/event-history.entity.js'
import { EventName } from '../../constants.js'
import { LoyaltyPointsDeptEntity } from '../../entitites/loyalty-points-dept.entity.js'

describe('webhook-events', () => {
  let app: FastifyInstance
  let db: Services
  let em: EntityManager

  beforeAll(async () => {
    [app, db] = await Promise.all([createApp(), initORM()])
    em = db.orm.em.fork();
    await app.ready()
  })

  beforeEach(async () => {
    await clearDatabase(db.orm)
  });

  afterAll(async () => {
    app.close()
  })

  it('should handle customer created event', async () => {
    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-13T21:03:10Z",
        "EventName": "CustomerCreated",
        "EntityName": "Customer",
        "Sequence": 1,
        "Payload": {
          "CustomerId": "814e496d-c6d1-49d5-b30b-359b4f83fa48"
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

  })

  it('should handle order placed event', async () => {
    const orderId = randomUUID()

    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-16T12:24:45+01:00",
        "EventName": "OrderPlaced",
        "EntityName": "Order",
        "Sequence": 10,
        "Payload": {
          "OrderId": orderId,
          "CustomerId": "814e496d-c6d1-49d5-b30b-359b4f83fa48",
          "TotalOrderAmount": 50
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    const pointsFromOrder = await em.findOne(LoyaltyPointsEntity, { orderId })
    expect(pointsFromOrder).toBeTruthy()
    expect(pointsFromOrder?.pointsFromOrder).toBe(1);
  })

  it('should handle order placed event when same event alredy occured', async () => {

    const eventName = EventName.ORDER_PLACED;
    const sequence = 10;

    const loyaltyPoints = new LoyaltyPointsEntity({
      customerId: randomUUID(),
      availablePoints: 1,
      pointsFromOrder: 1,
      orderId: randomUUID(),
      expiresAt: addDays(new Date(), 1),
    });

    const eventHistory = new EventsHistoryEntity({
      eventHash: hashEvent({ eventName, sequence }),
    });

    em.persist([loyaltyPoints, eventHistory]);
    await em.flush();

    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-16T12:24:45+01:00",
        "EventName": eventName,
        "EntityName": "Order",
        "Sequence": sequence,
        "Payload": {
          "OrderId": loyaltyPoints.orderId,
          "CustomerId": loyaltyPoints.customerId,
          "TotalOrderAmount": 50
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    const pointsFromOrder = await em.find(LoyaltyPointsEntity, { customerId: loyaltyPoints.customerId })
    expect(pointsFromOrder).toHaveLength(1);
  })

  it('should reduce debt balance when a customer with negative points places a new order', async () => {

    const eventName = EventName.ORDER_PLACED;
    const sequence = 10;
    const orderId = randomUUID();

    const loyaltyPointsDept = new LoyaltyPointsDeptEntity({
      customerId: randomUUID(),
      points: 1,
    });

    await em.persistAndFlush(loyaltyPointsDept);

    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-16T12:24:45+01:00",
        "EventName": eventName,
        "EntityName": "Order",
        "Sequence": sequence,
        "Payload": {
          "OrderId": orderId,
          "CustomerId": loyaltyPointsDept.customerId,
          "TotalOrderAmount": 100
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    const emFork = em.fork();

    const [pointsFromOrder, customerDept] = await Promise.all([
      emFork.findOne(LoyaltyPointsEntity, { customerId: loyaltyPointsDept.customerId }),
      emFork.findOne(LoyaltyPointsDeptEntity, { customerId: loyaltyPointsDept.customerId })
    ]);

    expect(pointsFromOrder).toBeTruthy();
    expect(pointsFromOrder?.pointsFromOrder).toBe(2);
    expect(pointsFromOrder?.availablePoints).toBe(1);

    expect(customerDept).toBeTruthy();
    expect(customerDept?.points).toBe(0);
  })

  it('should handle order canceled event when order does not exist', async () => {
    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-14T09:57:52+01:00",
        "EventName": "OrderCanceled",
        "EntityName": "Order",
        "Sequence": 2,
        "Payload": {
          "OrderId": "671eac98-7d4d-47fb-bd07-dee2c3a14d11"
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')
  })


  it('should handle order canceled event when order exist', async () => {
    const loyaltyPoints = new LoyaltyPointsEntity({
      customerId: randomUUID(),
      availablePoints: 10,
      pointsFromOrder: 10,
      orderId: randomUUID(),
      expiresAt: addDays(new Date(), 1),
    });

    await em.persistAndFlush(loyaltyPoints);

    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-14T09:57:52+01:00",
        "EventName": "OrderCanceled",
        "EntityName": "Order",
        "Sequence": 2,
        "Payload": {
          "OrderId": loyaltyPoints.orderId
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    const pointsFromOrder = await em.findOne(LoyaltyPointsEntity, { orderId: loyaltyPoints.orderId })
    expect(pointsFromOrder).toBeTruthy()
    expect(pointsFromOrder?.isCancelled).toBe(true);
  })

  it('should handle order returned event when order does not exist', async () => {
    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-11T16:54:43+01:00",
        "EventName": "OrderReturned",
        "EntityName": "Order",
        "Sequence": 2,
        "Payload": {
          "OrderId": "2e59361e-9544-11ee-a802-52f5db49a304"
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')
  })

  it('should handle order returned event when order exist', async () => {
    const loyaltyPoints = new LoyaltyPointsEntity({
      customerId: randomUUID(),
      availablePoints: 10,
      pointsFromOrder: 10,
      orderId: randomUUID(),
      expiresAt: addDays(new Date(), 1),
    });

    await em.persistAndFlush(loyaltyPoints);

    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-11T16:54:43+01:00",
        "EventName": "OrderReturned",
        "EntityName": "Order",
        "Sequence": 2,
        "Payload": {
          "OrderId": loyaltyPoints.orderId
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    const pointsFromOrder = await em.fork().findOne(LoyaltyPointsEntity, { orderId: loyaltyPoints.orderId })
    expect(pointsFromOrder).toBeTruthy()
    expect(pointsFromOrder?.isCancelled).toBe(true);
  })

  it('should create dept when customer consumed points from order which is canceled and does not have any points left', async () => {
    const loyaltyPoints = new LoyaltyPointsEntity({
      customerId: randomUUID(),
      availablePoints: 2,
      pointsFromOrder: 10,
      orderId: randomUUID(),
      expiresAt: addDays(new Date(), 1),
    });

    await em.persistAndFlush(loyaltyPoints);

    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-11T16:54:43+01:00",
        "EventName": "OrderReturned",
        "EntityName": "Order",
        "Sequence": 2,
        "Payload": {
          "OrderId": loyaltyPoints.orderId
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    const emFork = em.fork();

    const [pointsDept, pointsFromOrder] = await Promise.all([
      emFork.findOne(LoyaltyPointsDeptEntity, { customerId: loyaltyPoints.customerId }),
      emFork.findOne(LoyaltyPointsEntity, { orderId: loyaltyPoints.orderId })
    ])

    expect(pointsFromOrder).toBeTruthy()
    expect(pointsFromOrder?.isCancelled).toBe(true);
    expect(pointsDept).toBeTruthy();
    expect(pointsDept?.points).toBe(8);
  })

  it('should prioritize available points when canceling an order with previously consumed points', async () => {
    const customerId = randomUUID();

    const loyaltyPoints = [new LoyaltyPointsEntity({
      customerId,
      availablePoints: 2,
      pointsFromOrder: 10,
      orderId: randomUUID(),
      expiresAt: addDays(new Date(), 1),
    }),
    new LoyaltyPointsEntity({
      customerId,
      availablePoints: 8,
      pointsFromOrder: 8,
      orderId: randomUUID(),
      expiresAt: addDays(new Date(), 1),
    })
    ];

    await em.persistAndFlush(loyaltyPoints);

    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-11T16:54:43+01:00",
        "EventName": "OrderReturned",
        "EntityName": "Order",
        "Sequence": 2,
        "Payload": {
          "OrderId": loyaltyPoints[0].orderId
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    const emFork = em.fork();

    const [pointsDept, pointsFromCanceledOrder, pointsFromOtherOrder] = await Promise.all([
      emFork.findOne(LoyaltyPointsDeptEntity, { customerId }),
      emFork.findOne(LoyaltyPointsEntity, { orderId: loyaltyPoints[0].orderId }),
      emFork.findOne(LoyaltyPointsEntity, { orderId: loyaltyPoints[1].orderId })
    ])

    expect(pointsFromCanceledOrder).toBeTruthy()
    expect(pointsFromCanceledOrder?.isCancelled).toBe(true);
    expect(pointsFromOtherOrder).toBeTruthy();
    expect(pointsFromOtherOrder?.availablePoints).toBe(0);
    expect(pointsDept).toBeFalsy();
  })

  it('should create dept when customer consumed points from order which is canceled and does not have enough points left', async () => {
    const customerId = randomUUID();

    const loyaltyPoints = [new LoyaltyPointsEntity({
      customerId,
      availablePoints: 2,
      pointsFromOrder: 10,
      orderId: randomUUID(),
      expiresAt: addDays(new Date(), 1),
    }),
    new LoyaltyPointsEntity({
      customerId,
      availablePoints: 2,
      pointsFromOrder: 2,
      orderId: randomUUID(),
      expiresAt: addDays(new Date(), 1),
    })
    ];

    await em.persistAndFlush(loyaltyPoints);

    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-11T16:54:43+01:00",
        "EventName": "OrderReturned",
        "EntityName": "Order",
        "Sequence": 2,
        "Payload": {
          "OrderId": loyaltyPoints[0].orderId
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    const emFork = em.fork();

    const [pointsDept, pointsFromCanceledOrder, pointsFromOtherOrder] = await Promise.all([
      emFork.findOne(LoyaltyPointsDeptEntity, { customerId }),
      emFork.findOne(LoyaltyPointsEntity, { orderId: loyaltyPoints[0].orderId }),
      emFork.findOne(LoyaltyPointsEntity, { orderId: loyaltyPoints[1].orderId })
    ])

    expect(pointsFromCanceledOrder).toBeTruthy()
    expect(pointsFromCanceledOrder?.isCancelled).toBe(true);
    expect(pointsFromOtherOrder).toBeTruthy();
    expect(pointsFromOtherOrder?.availablePoints).toBe(0);
    expect(pointsDept).toBeTruthy();
    expect(pointsDept?.points).toBe(6);
  })

  it('should handle customer delete event when customer does not have any orders', async () => {
    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-20T00:57:57+01:00",
        "EventName": "CustomerDeleted",
        "EntityName": "Customer",
        "Sequence": 2,
        "Payload": {
          "CustomerId": "814e496d-c6d1-49d5-b30b-359b4f83fa48"
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')
  })

  it('should delete user dept when customer deleted occured', async () => {

    const loyaltyPointsDept = new LoyaltyPointsDeptEntity({
      customerId: randomUUID(),
      points: 10,
    });

    await em.persistAndFlush(loyaltyPointsDept);

    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-20T00:57:57+01:00",
        "EventName": "CustomerDeleted",
        "EntityName": "Customer",
        "Sequence": 2,
        "Payload": {
          "CustomerId": loyaltyPointsDept.customerId
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    const pointsDept = await em.fork().findOne(LoyaltyPointsDeptEntity, { customerId: loyaltyPointsDept.customerId })
    expect(pointsDept).toBeFalsy();
  })

  it('should delete user loyalty points when customer deleted occured', async () => {
    const loyaltyPoints = new LoyaltyPointsEntity({
      customerId: randomUUID(),
      availablePoints: 2,
      pointsFromOrder: 2,
      orderId: randomUUID(),
      expiresAt: addDays(new Date(), 1),
    });

    await em.persistAndFlush(loyaltyPoints);

    await supertest(app.server)
      .post(`/webhook`)
      .send({
        "EventTime": "2023-11-20T00:57:57+01:00",
        "EventName": "CustomerDeleted",
        "EntityName": "Customer",
        "Sequence": 2,
        "Payload": {
          "CustomerId": loyaltyPoints.customerId
        }
      })
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    const pointsDept = await em.fork().findOne(LoyaltyPointsEntity, { customerId: loyaltyPoints.customerId })
    expect(pointsDept).toBeFalsy();
  })
})
