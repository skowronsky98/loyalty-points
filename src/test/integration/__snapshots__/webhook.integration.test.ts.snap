// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`webhook > snapshot 1`] = `
{
  "foo": "bar",
}
`;
